Commit ID: ad2f27a2a3d26a0d4fb7044044cc09cf4ec57288
Change ID: lqnrryqyluwrrupuutpwnvksxtyqwrss
Bookmarks: mixed-ngon-fix mixed-ngon-fix@git mixed-ngon-fix@me
Author   : Ashley Ruglys <ashley.ruglys@gmail.com> (2026-02-20 19:54:39)
Committer: Ashley Ruglys <ashley.ruglys@gmail.com> (2026-02-20 19:56:43)

    Accumulate vertex offset correctly for objects with differing face vertex counts

diff --git a/source/xatlas/xatlas.cpp b/source/xatlas/xatlas.cpp
index 5c5c57ecab..03c2a6da0a 100644
--- a/source/xatlas/xatlas.cpp
+++ b/source/xatlas/xatlas.cpp
@@ -9087,13 +9087,14 @@
 	uint32_t warningCount = 0;
 	internal::Array<uint32_t> triIndices;
 	internal::Triangulator triangulator;
+	uint32_t vertexOffset = 0;
 	for (uint32_t face = 0; face < faceCount; face++) {
 		// Decode face indices.
 		const uint32_t faceVertexCount = meshDecl.faceVertexCount ? (uint32_t)meshDecl.faceVertexCount[face] : 3;
 		uint32_t polygon[UINT8_MAX];
 		for (uint32_t i = 0; i < faceVertexCount; i++) {
 			if (hasIndices) {
-				polygon[i] = DecodeIndex(meshDecl.indexFormat, meshDecl.indexData, meshDecl.indexOffset, face * faceVertexCount + i);
+				polygon[i] = DecodeIndex(meshDecl.indexFormat, meshDecl.indexData, meshDecl.indexOffset, vertexOffset + i);
 				// Check if any index is out of range.
 				if (polygon[i] >= meshDecl.vertexCount) {
 					mesh->~Mesh();
@@ -9101,9 +9102,10 @@
 					return AddMeshError::IndexOutOfRange;
 				}
 			} else {
-				polygon[i] = face * faceVertexCount + i;
+				polygon[i] = vertexOffset + i;
 			}
 		}
+		vertexOffset += faceVertexCount;
 		// Ignore faces with degenerate or zero length edges.
 		bool ignore = false;
 		for (uint32_t i = 0; i < faceVertexCount; i++) {
